tao_compiler/mlir/disc/utils/source_emitter.cc

```diff
@@ -0,0 +1,590 @@
+  std::string type_str = MLIRType2CUDATypeStr(result_type);
+
+  std::string result_name = EmitUniqueName("dyn_bcast_indim");
+
+  assert(binding.count(result) == 0);
+  binding[result] = result_name;
+
+  return type_str + " " + result_name + " = " + expression;
+}
+
+llvm::Optional<std::string> SourceEmitterCUDA::EmitOp(
+    Operation* op, ValueNameBinding& binding) {
+  if (isa<lmhlo::AbsOp, lmhlo::CeilOp, lmhlo::ConvertOp, lmhlo::CosineOp,
+          lmhlo::ExpOp, lmhlo::FloorOp, lmhlo::IsFiniteOp, lmhlo::LogOp,
+          lmhlo::Log1pOp, lmhlo::NegOp, lmhlo::NotOp, lmhlo::RsqrtOp,
+          lmhlo::SqrtOp, lmhlo::TanhOp>(op)) {
+    return EmitElemWiseUnaryOp(op, binding);
+  } else if (isa<lmhlo::AddOp, lmhlo::SubtractOp, lmhlo::MulOp, lmhlo::DivOp,
+                 lmhlo::MaxOp, lmhlo::MinOp, lmhlo::CompareOp, lmhlo::AndOp,
+                 lmhlo::OrOp, lmhlo::RemOp, lmhlo::PowOp>(op)) {
+    return EmitElemWiseBinaryOp(op, binding);
+  } else if (isa<lmhlo::SelectOp, lmhlo::ClampOp>(op)) {
+    return EmitElemWiseTernaryOp(op, binding);
+  } else if (isa<lmhlo::ConstantOp>(op)) {
+    return EmitScalarOrSplatConstantOp(op, binding);
+  } else if (isa<lmhlo::DynamicBroadcastInDimOp>(op)) {
+    return EmitBroadcastOfScalarOrSplatConstantOp(op, binding);
+  } else {
+    return llvm::None;
+  }
+}
+
+std::string SourceEmitterCUDA::EmitUniqueName(llvm::StringRef op_str) {
+  std::string name = "bladedisc_" + op_str.str();
+  if (existing_names_.count(op_str.str()) == 0) {
+    existing_names_.try_emplace(op_str.str(), 0);
+  }
+  int32_t count = existing_names_[op_str.str()]++;
+  name += "_" + to_string(count);
+
+  return name;
+}
+
+void SourceEmitterCUDA::bindValueNames(
+    const SmallVectorImpl<Value>& inputs,
+    const SmallVectorImpl<std::string>& names, ValueNameBinding& binding) {
+  assert(inputs.size() == names.size());
+  for (int64_t i = 0; i < inputs.size(); i++) {
+    binding[inputs[i]] = names[i];
+  }
+}
+
+bool SourceEmitterCUDA::isSupportedOp(Operation* op) {
+  if (isa<lmhlo::AbsOp, lmhlo::CeilOp, lmhlo::ConvertOp, lmhlo::CosineOp,
+          lmhlo::ExpOp, lmhlo::FloorOp, lmhlo::IsFiniteOp, lmhlo::LogOp,
+          lmhlo::Log1pOp, lmhlo::NegOp, lmhlo::NotOp, lmhlo::RsqrtOp,
+          lmhlo::SqrtOp, lmhlo::TanhOp>(op) ||
+      isa<lmhlo::AddOp, lmhlo::SubtractOp, lmhlo::MulOp, lmhlo::DivOp,
+          lmhlo::MaxOp, lmhlo::MinOp, lmhlo::CompareOp, lmhlo::AndOp,
+          lmhlo::OrOp, lmhlo::RemOp, lmhlo::PowOp>(op) ||
+      isa<lmhlo::SelectOp, lmhlo::ClampOp>(op)) {
+    return true;
+  } else if (isa<lmhlo::ConstantOp>(op)) {
+    lmhlo::ConstantOp constant = dyn_cast<lmhlo::ConstantOp>(op);
+    MemRefType memref_type = constant.getOutput().getType().cast<MemRefType>();
+    return memref_type.getRank() == 0 || constant.getValue().isSplat();
+  } else if (isa<lmhlo::DynamicBroadcastInDimOp>(op)) {
+    // Only deal with the case that the last rewriter in the same block is
```

Comment:
Why such limitation? Seems like BiasAdd doesn't meet this requirement.

Reply:
The constraint is for the first version currently. After iterative enhancements, a more general `broadcast` will be supported.

Reply:
> 'such limitation' means scalar constant ?

Yes. Scalar or splat-constant.